airflow/api_fastapi/core_api/security.py

```diff
@@ -75,23 +83,63 @@ async def get_user_with_exception_handling(request: Request) -> BaseUser | None:
     return get_user(token_str)
 
 
-def requires_access_dag(method: ResourceMethod, access_entity: DagAccessEntity | None = None) -> Callable:
+def requires_access_dag(
+    method: ResourceMethod, access_entity: DagAccessEntity | None = None
+) -> Callable[[Request, BaseUser | None], None]:
     def inner(
-        dag_id: str | None = None,
+        request: Request,
         user: Annotated[BaseUser | None, Depends(get_user)] = None,
     ) -> None:
-        def callback():
-            return get_auth_manager().is_authorized_dag(
-                method=method, access_entity=access_entity, details=DagDetails(id=dag_id), user=user
-            )
+        dag_id: str | None = request.path_params.get("dag_id")
 
         _requires_access(
-            is_authorized_callback=callback,
+            is_authorized_callback=lambda: get_auth_manager().is_authorized_dag(
+                method=method, access_entity=access_entity, details=DagDetails(id=dag_id), user=user
+            )
         )
 
     return inner
 
 
+class PermittedDagFilter(BaseParam[set[str]]):
```

Comment:
Just an idea.

I don't think those should be a `BaseParam`. They are not query parameters or path parameters.

Nonetheless it is important for them to implement the `to_orm` interface. We could define a Protocol and both `BaseParam` and These permissions type could implement. Protocol typing is a little weird, so maybe just a baseclass. `HasMappingRelation`, `ORMClause` or any other name, this onees are not great, that just needs a `to_orm` method and maybe a `self.value`.

And then `paginated_select` takes `Sequence[ORMClause]`.

This will allow to not have `class PermittedDagFilter(BaseParam[set[str]]):` but just `class PermittedDagFilter(ORMClause[set[str]]):` and not confuse reader that it is a query param or path parameter. And then `class BaseParam(ORMClause, ABC):`